### PR TITLE
DAOS-2263 pool: Store owner/group in pool props

### PR DESCRIPTION
**Pull request by kjacque**
Created two new pool properties, Owner and Owner-Group. These will eventually be a replacement for the uid/gid/mode structure currently stored in pool properties, and be set at pool creation time. For now, they are unused.

The owner and owner-group are strings in the format "user@[domain]". At this time the domain is optional.

### CODE REVIEW

File: src/common/misc.c [hunk 1]
```diff
@@ -664,6 +664,8 @@ daos_prop_free(daos_prop_t *prop)
 		switch (entry->dpe_type) {
 		case DAOS_PROP_PO_LABEL:
 		case DAOS_PROP_CO_LABEL:
+		case DAOS_PROP_PO_OWNER:
+		case DAOS_PROP_PO_OWNER_GROUP:
```
Comment (liuxuezhao) [inline src/common/misc.c]: just curious why the OWNER, GROUP stored by a string rather than uid/gid?

File: src/pool/srv_pool.c [hunk 2]
```diff
@@ -1410,6 +1439,47 @@ pool_prop_read(struct rdb_tx *tx, const struct pool_svc *svc, uint64_t bits,
 		       value.iov_buf_len);
 		idx++;
 	}
+	if (bits & DAOS_PO_QUERY_PROP_OWNER) {
+		daos_iov_set(&value, NULL, 0);
+		rc = rdb_tx_lookup(tx, &svc->ps_root, &ds_pool_prop_owner,
+				   &value);
+		if (rc != 0)
+			return rc;
+		/* IOV len includes strlen + null terminator */
```
Comment (liw) [inline src/pool/srv_pool.c]: Looks like the owner and owner_group values written by pool_prop_write do not include the '\0'. Or, did I misread the code?

File: src/pool/srv_pool.c [hunk 3]
```diff
@@ -1410,6 +1439,47 @@ pool_prop_read(struct rdb_tx *tx, const struct pool_svc *svc, uint64_t bits,
...
+			return -DER_NOMEM;
+		idx++;
+	}
+	if (bits & DAOS_PO_QUERY_PROP_OWNER_GROUP) {
+		daos_iov_set(&value, NULL, 0);
+		rc = rdb_tx_lookup(tx, &svc->ps_root, &ds_pool_prop_owner_group,
+				   &value);
+		if (rc != 0)
+			return rc;
+		/* IOV len includes strlen + null terminator */
```
Comment (liw) [inline src/pool/srv_pool.c]: Ditto.

File: src/common/misc.c (hunk 1 above)
Comment (kjacque) [reply src/common/misc.c]: It has to do with our future goals for access control. Right now it's just Unix users/groups, but we'd like to support other authentication mechanisms in the future (Kerberos has come up as an example) which may not map directly to one specific uid/gid.

File: src/pool/srv_pool.c (hunk 2 above)
Comment (kjacque) [reply src/pool/srv_pool.c]: Good catch. Will fix.

File: src/pool/srv_pool.c (hunk 3 above)
Comment (kjacque) [reply src/pool/srv_pool.c]: Fixed

File: src/tests/suite/daos_pool.c [hunk 4]
```diff
@@ -491,7 +491,7 @@ pool_properties(void **state)
 	rc = daos_pool_query(arg->pool.poh, NULL, NULL, prop_query, NULL);
 	assert_int_equal(rc, 0);
 
-	assert_int_equal(prop_query->dpp_nr, 5);
+	assert_int_equal(prop_query->dpp_nr, 7);
```
Comment (dpquigl) [inline src/tests/suite/daos_pool.c]: I know these magic numbers were already there and you were incrementing them but is there a define we could use instead somewhere?

Comment (kjacque) [reply src/tests/suite/daos_pool.c]: I can move the define that's in pool/srv_layout.c.

File: src/common/misc.c [hunk 5]
```diff
@@ -872,6 +924,17 @@ daos_prop_dup(daos_prop_t *prop, bool pool)
 		case DAOS_PROP_CO_ACL:
 			/* TODO: Implement container ACL */
 			break;
+		case DAOS_PROP_PO_OWNER:
+		case DAOS_PROP_PO_OWNER_GROUP:
+			entry_dup->dpe_str =
+					strndup(entry->dpe_str,
```
[the remark this answers is not in the document]
Comment (kjacque) [reply src/common/misc.c]: Fixed

File: src/common/misc.c [hunk 6]
```diff
@@ -960,6 +1025,20 @@ daos_prop_copy(daos_prop_t *prop_req, daos_prop_t *prop_reply)
 			acl_alloc = entry_req->dpe_val_ptr;
 		} else if (type == DAOS_PROP_CO_ACL) {
 			/* TODO: Implement container ACL */
+		} else if (type == DAOS_PROP_PO_OWNER) {
+			entry_req->dpe_str =
+					strndup(entry_reply->dpe_str,
```
[the remark this answers is not in the document]
Comment (kjacque) [reply src/common/misc.c]: Fixed

File: src/common/misc.c [hunk 7]
```diff
@@ -960,6 +1025,20 @@ daos_prop_copy(daos_prop_t *prop_req, daos_prop_t *prop_reply)
...
+						DAOS_ACL_MAX_PRINCIPAL_LEN);
+			if (entry_req->dpe_str == NULL)
+				D_GOTO(out, rc = -DER_NOMEM);
+			owner_alloc = entry_req->dpe_str;
+		} else if (type == DAOS_PROP_PO_OWNER_GROUP) {
+			entry_req->dpe_str =
+					strndup(entry_reply->dpe_str,
```
[the remark this answers is not in the document]
Comment (kjacque) [reply src/common/misc.c]: Fixed

File: src/common/misc.c [hunk 8]
```diff
@@ -914,6 +977,8 @@ daos_prop_copy(daos_prop_t *prop_req, daos_prop_t *prop_reply)
 	struct daos_prop_entry	*entries_alloc = NULL;
 	d_string_t		 label_alloc = NULL;
 	void			*acl_alloc = NULL;
+	d_string_t		 owner_alloc = NULL;
+	d_string_t		 group_alloc = NULL;
```
[the remark this answers is not in the document]
Comment (kjacque) [reply src/common/misc.c]: Good to know. Thanks for the explanation.

File: src/common/misc.c [hunk 9]
```diff
@@ -681,6 +683,46 @@ daos_prop_free(daos_prop_t *prop)
...
+static bool
+daos_prop_str_valid(d_string_t str, const char *prop_name, size_t max_len)
+{
+	size_t len;
+
+	if (str == NULL) {
+		D_ERROR("invalid NULL %s\n", prop_name);
+		return false;
+	}
+	len = strlen(str);
```
Comment (dpquigl) [inline src/common/misc.c]: Just noticed this. Are we sure that str will always be null terminated? If not we should use a strnlen here (really we should probably use it regardless).

Comment (jolivier23) [reply src/common/misc.c]: I've seen codacy complain about strnlen as well but yeah, it's good practice, particularly if there is a known maximum length to use strnlen.

Comment (kjacque) [reply src/common/misc.c]: Everywhere daos_prop_valid is used on an internally-generated string, we know the string is null terminated. But it's used on user inputs as well. I'll change it to the safer strnlen.

**Pull request merged**